tests/unittests/Foundation/NSUndoManagerTests.mm

```diff
@@ -0,0 +1,161 @@
+
+    EXPECT_EQ(object.value.intValue, oldValue.intValue);
+    [object updateNumber:newValue];
+    EXPECT_EQ([object.undoManager canUndo], YES);
+    EXPECT_EQ(object.value.intValue, newValue.intValue);
+
+    [object undo];
+    EXPECT_EQ(object.value.intValue, oldValue.intValue);
+    EXPECT_EQ([object.undoManager canUndo], NO);
+}
+
+TEST(NSUndoManager, undoTests) {
+    NSNumber* val1 = [NSNumber numberWithInt:5];
+    NSNumber* val2 = [NSNumber numberWithInt:6];
+    NSNumber* val3 = [NSNumber numberWithInt:7];
+    NSNumber* val4 = [NSNumber numberWithInt:8];
+
+    someObject* object = [[[someObject alloc] initWithValue:val1] autorelease];
+    EXPECT_EQ([object.undoManager canUndo], NO);
+
+    // Each undo operation here gets added to the single default undo group.
+    [object updateNumber:val2];
+    [object updateNumber:val3];
+    [object updateNumber:val4];
+    EXPECT_EQ(object.value.intValue, val4.intValue);
+    [object undo];
+    EXPECT_EQ(object.value.intValue, val1.intValue);
+    EXPECT_EQ([object.undoManager canUndo], NO);
+}
+
+TEST(NSUndoManager, undoRedoTests) {
+    NSNumber* val1 = [NSNumber numberWithInt:5];
+    NSNumber* val2 = [NSNumber numberWithInt:6];
+    NSNumber* val3 = [NSNumber numberWithInt:7];
+    NSNumber* val4 = [NSNumber numberWithInt:8];
+
+    someObject* object = [[[someObject alloc] initWithValue:val1] autorelease];
+    EXPECT_EQ([object.undoManager canUndo], NO);
+
+    [object updateNumber:val2];
+    [object updateNumber:val3];
+    [object updateNumber:val4];
+    EXPECT_EQ([object.undoManager canUndo], YES);
+    EXPECT_EQ(object.value.intValue, val4.intValue);
+    [object undo];
+    EXPECT_EQ([object.undoManager canRedo], YES);
+    EXPECT_EQ(object.value.intValue, val1.intValue);
+    [object redo];
+    EXPECT_EQ(object.value.intValue, val4.intValue);
+    EXPECT_EQ([object.undoManager canRedo], NO);
+    [object undo];
+    EXPECT_EQ(object.value.intValue, val1.intValue);
+    EXPECT_EQ([object.undoManager canRedo], YES);
+}
+
+TEST(NSUndoManager, undoRedoTestsWithMutableObject) {
+    NSNumber* val1 = [NSNumber numberWithInt:5];
+    NSNumber* val2 = [NSNumber numberWithInt:6];
+    NSNumber* val3 = [NSNumber numberWithInt:7];
+    NSNumber* val4 = [NSNumber numberWithInt:8];
+
+    someObject* object = [[[someObject alloc] initWithEmptyArray] autorelease];
+    EXPECT_EQ([object.undoManager canUndo], NO);
+
+    NSMutableArray* comparisonArray = [NSMutableArray array];
+    [comparisonArray addObject:val1];
+    [comparisonArray addObject:val2];
+    [comparisonArray addObject:val3];
+
+    [object addToMutableArray:val1];
+    [object addToMutableArray:val2];
+    [object addToMutableArray:val3];
+
+    EXPECT_OBJCEQ(comparisonArray, object.arrayOfNumbers);
+
+    [object undo];
+
+    EXPECT_OBJCEQ([NSMutableArray array], object.arrayOfNumbers);
+
+    [object redo];
+
+    EXPECT_OBJCEQ(comparisonArray, object.arrayOfNumbers);
+}
```

Comment:
should probably add tests for trying to undo/redo when there's nothing to undo/redo to guarantee the objects aren't modified